File: 0096-unique-binary-search-trees/0096-unique-binary-search-trees.py
class Solution:
    def numTrees(self, n: int) -> int:
        dp = [1] * (n+1)
        for i in range(2, n+1):
            total = 0
            for j in range(1, i+1):
                left = j - 1
                right = i - j
                total += dp[left] * dp[right]
            dp[i] = total
        return dp[n]

File: 0096-unique-binary-search-trees/0096-unique-binary-search-trees.js
/**
 * @param {number} n
 * @return {number}
 */
var numTrees = function(n) {
    const dp = Array(n+1).fill(1)
    for (let numNodes=2; numNodes <= n; numNodes++) {
        let total = 0
        for (let root=1; root <= numNodes; root++) {
            const left = root-1
            const right = numNodes - root
            total += dp[left] * dp[right]
        }
        dp[numNodes] = total
    }
    return dp[n]
};